packages/core/src/RenderPipeline/MaskManager.ts

```diff
@@ -19,29 +25,109 @@ export class MaskManager {
     return (MaskManager._maskDecrementRenderQueue ||= new RenderQueue(RenderQueueType.Transparent));
   }
 
-  allSpriteMasks = new DisorderedArray<SpriteMask>();
-  preMaskLayer = 0;
+  notWriteStencil = true;
+  private _preMaskLayer = SpriteMaskLayer.Nothing;
+  private _allSpriteMasks = new DisorderedArray<SpriteMask>();
 
   addSpriteMask(mask: SpriteMask): void {
-    mask._maskIndex = this.allSpriteMasks.length;
-    this.allSpriteMasks.add(mask);
+    mask._maskIndex = this._allSpriteMasks.length;
+    this._allSpriteMasks.add(mask);
   }
 
   removeSpriteMask(mask: SpriteMask): void {
-    const replaced = this.allSpriteMasks.deleteByIndex(mask._maskIndex);
+    const replaced = this._allSpriteMasks.deleteByIndex(mask._maskIndex);
     replaced && (replaced._maskIndex = mask._maskIndex);
     mask._maskIndex = -1;
   }
 
-  buildMaskRenderElement(
-    element: SubRenderElement,
+  drawMask(context: RenderContext, pipelineStageTagValue: string, maskLayer: SpriteMaskLayer): void {
+    const incrementMaskQueue = MaskManager.getMaskIncrementRenderQueue();
+    incrementMaskQueue.clear();
+    const decrementMaskQueue = MaskManager.getMaskDecrementRenderQueue();
+    decrementMaskQueue.clear();
+
+    this._buildMaskRenderElement(maskLayer, incrementMaskQueue, decrementMaskQueue);
+
+    const engine = context.camera.engine;
+    const batcherManager = engine._batcherManager;
+    incrementMaskQueue.batch(batcherManager);
+    batcherManager.uploadBuffer();
+    incrementMaskQueue.render(context, pipelineStageTagValue, RenderQueueMaskType.Increment);
+    decrementMaskQueue.batch(batcherManager);
+    batcherManager.uploadBuffer();
+    decrementMaskQueue.render(context, pipelineStageTagValue, RenderQueueMaskType.Decrement);
+  }
+
+  clearMask(context: RenderContext, pipelineStageTagValue: string): void {
+    const preMaskLayer = this._preMaskLayer;
+    if (preMaskLayer !== SpriteMaskLayer.Nothing) {
+      if (this.notWriteStencil) {
+        const engine = context.camera.engine;
+        engine._hardwareRenderer.clearRenderTarget(engine, CameraClearFlags.Stencil, null);
+      } else {
+        const decrementMaskQueue = MaskManager.getMaskDecrementRenderQueue();
+        decrementMaskQueue.clear();
+
+        const masks = this._allSpriteMasks;
+        const maskElements = masks._elements;
+        for (let i = 0, n = masks.length; i < n; i++) {
+          const mask = maskElements[i];
+          mask.influenceLayers & preMaskLayer && decrementMaskQueue.pushRenderElement(mask._renderElement);
+        }
+
+        const batcherManager = context.camera.engine._batcherManager;
+        decrementMaskQueue.batch(batcherManager);
+        batcherManager.uploadBuffer();
+        decrementMaskQueue.render(context, pipelineStageTagValue, RenderQueueMaskType.Decrement);
+      }
+
+      this._preMaskLayer = SpriteMaskLayer.Nothing;
+    }
+  }
+
+  checkStencilAccess(material: Material): StencilAccess {
```

Comment:
Two choices:
- check Stencil write is enough
- If you want improve, i think you can revert stencil created by mask when custom material want to read!

Reply:
done